sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientWrapper.scala

```diff
@@ -62,6 +63,39 @@ private[hive] class ClientWrapper(
   extends ClientInterface
   with Logging {
 
+  overrideHadoopShims()
+
+  // !! HACK ALERT !!
+  //
+  // This method is used to workaround CDH Hadoop versions like 2.0.0-mr1-cdh4.1.1.
+  //
+  // Internally, Hive `ShimLoader` tries to load different versions of Hadoop shims by checking
+  // version information gathered from Hadoop jar files.  If the major version number is 1,
+  // `Hadoop20SShims` will be loaded.  Otherwise, if the major version number is 2, `Hadoop23Shims`
+  // will be chosen.  However, CDH Hadoop versions like 2.0.0-mr1-cdh4.1.1 have 2 as major version
+  // number, but contain Hadoop 1 code.  This confuses Hive `ShimLoader` and loads wrong version of
```

Comment:
@vanzin Thanks for the explanation, then I guess it would be sufficient to special case 2.0.x and 2.1.x CDH versions. (Or dropping CDH4 support as you mentioned offline.)

@srowen I've updated the comment to reflect the facts (hopefully).